control/dns_control.go

```diff
@@ -769,3 +845,74 @@ func (c *DnsController) dialSend(invokingDepth int, req *udpRequest, data []byte
 	}
 	return nil
 }
+
+func httpDNS(client *http.Client, target string, data []byte) (respMsg *dnsmessage.Msg, err error) {
+	serverURL := url.URL{
+		Scheme: "https",
+		Host:   target,
+		Path:   "/dns-query",
+	}
+
+	req, err := http.NewRequest(http.MethodPost, serverURL.String(), strings.NewReader(string(data)))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/dns-message")
+	req.Header.Set("Accept", "application/dns-message")
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	buf, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	var msg dnsmessage.Msg
+	if err = msg.Unpack(buf); err != nil {
+		return nil, err
+	}
+	respMsg = &msg
+	return respMsg, nil
+}
+
+type stream interface {
+	io.Reader
+	io.Writer
+}
+
+func streamDNS(stream stream, data []byte) (respMsg *dnsmessage.Msg, err error) {
```

Comment:
This function can be divided into sending and receiving. Add some comments to make the division clearer. Alternatively, you can change it to sendStreamDNS.